security_token.py

```diff
@@ -208,40 +208,55 @@ def getRuleKey(sendGroup, receiveGroup):
+        return Not(Or(
+            Lt(App.localGet(sender_idx, Bytes("balance")), amount), # check sender balance
+            App.globalGet(Bytes("paused")), # can't transfer when the contract is paused
+            App.localGet(sender_idx, Bytes("frozen")), # sender account can't be frozen
+            App.localGet(sender_idx, Bytes("lockUntil")) >= Global.latest_timestamp(), # sender account can't be locked
+
+            # check that a transfer rule exists and allows the transfer from the sender to the receiver at the current time
+            App.globalGet(getRuleKey(App.localGet(sender_idx, Bytes("transferGroup")), App.localGet(receiver_idx, Bytes("transferGroup")))) < Int(1),
+            App.globalGet(getRuleKey(App.localGet(sender_idx, Bytes("transferGroup")), App.localGet(receiver_idx, Bytes("transferGroup")))) >= Global.latest_timestamp(),
+
+            # check that max balance is not exceeded
+            And(
+                App.localGet(receiver_idx, Bytes("maxBalance")),
+                App.localGet(receiver_idx, Bytes("maxBalance")) < App.localGet(receiver_idx, Bytes("balance")) + amount
+            )
+        ))
+
+    # detectTransferRestriction
+    # transaction succeeds if transfer is possible from the sender Txn.accounts[1] to the receiver Txn.accounts[2]
+    # goal app call --app-id uint --from address --app-account senderAddr --app-account receiverAddr --app-arg 'str:detectTransferRestriction' --app-arg "int:amount"
+    # checks are made to see if the sender account is frozen or locked
+    # checks are made to see if there is a transfer rule allowing transfer between sender and receiver transfer groups
+    # the transfer must occur after the transfer group lockUntil date
+    transfer_amount = Btoi(Txn.application_args[1])
+    detectTransferRestriction = Seq([
+        Assert(And(
+            Txn.application_args.length() == Int(2),
+            Txn.accounts.length() == Int(2),
+            isTransferAllowed(Int(1), Int(2), transfer_amount)
+        )),
+        Return(Int(1))
+    ])
+
     # transfer
     # transfers assets from the sender to the receiver Txn.accounts[1]
     # goal app call --app-id uint --from address --app-account receiverAddr --app-arg 'str:transfer' --app-arg "int:amount"
     # checks are made to see if the sender account is frozen or locked
     # checks are made to see if there is a transfer rule allowing transfer between sender and receiver transfer groups
     # the transfer must occur after the transfer group lockUntil date
     transfer_amount = Btoi(Txn.application_args[1])
-    receiver_max_balance = App.localGetEx(Int(1), App.id(), Bytes("maxBalance"))
     transfer = Seq([
         Assert(And(
             Txn.application_args.length() == Int(2),
             Txn.accounts.length() == Int(1),
-            transfer_amount <= App.localGet(Int(0), Bytes("balance"))
+            isTransferAllowed(Int(0), Int(1), transfer_amount)
```

Comment:
@Alexey1100 

1) I think one way to reduce the byte size to < 1000 bytes would be to merge the detectTransferRestriction and transfer code.

Based on which "function" was called in arg-1 we could switch between:
* "detectTransferRestriction" which would check the addresses passed in (not the sender address) and *would not update balances*
* "transfer" which would use the sender from address and would update balances

2) I think this approach is better than reducing the size of the Byte("strings"). I tried reducing function name strings down size to 1 character and it didn't drop the size below 1000 bytes.

3) We also need to add tests.

Reply:
> to merge the detectTransferRestriction and transfer code.

@Alexey1100 I think it will lead to the same issue as described here https://github.com/CoMakery/algorand-security-token/pull/17#discussion_r568298358.
Since `Txn.accounts[]` are different for each call TEAL will duplicate the logic.